File: Java/src/main/java/QuestionBank/Solution501_600/Solution563.java
package QuestionBank.Solution501_600;

import tool.TreeNode;

/**
 * @author windows
 */
public class Solution563 {
    public int findTilt(TreeNode root) {
        if (root == null){
            return 0;
        }
        int left = findTilt(root.left);
        int right = findTilt(root.right);
        int leftVal = root.left == null ? 0 : findTiltVal(root.left);
        int rightVal = root.right == null ? 0 : findTiltVal(root.right);
        return Math.abs(rightVal - leftVal) + left + right;
    }

    private int findTiltVal(TreeNode root) {
        if (root == null){
            return 0;
        }
        int left = findTiltVal(root.left);
        int right = findTiltVal(root.right);
        return left + right + root.val;
    }


    public static void main(String[] args) {
        Solution563 solution563 = new Solution563();
    }
}

File: C++/src/Solution501_600/Solution563.cpp
//
// Created by windows on 2023/3/2.
//

#include "Solution563.h"
#include <functional>
#include <iostream>
#include <vector>
#include <unordered_set>
#include <unordered_map>
#include <string>
using namespace std;

//Definition for a binary tree node.
struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};



int findTilt(TreeNode* root) {
  int res = 0;

  function<int(TreeNode*)> dfs = [&](TreeNode* root){
    if(!root) return 0;
    int left = dfs(root->left);
    int right = dfs(root->right);
    res += abs(left - right);
    return left + right + root->val;
  };
  dfs(root);
  return res;
}